testers.lycheeLinkCheck: init (#298665)
NixOS/nixpkgs: doc/build-helpers/testers.chapter.md

:::

## `lycheeLinkCheck` {#tester-lycheeLinkCheck}

Check a packaged static site's links with the [`lychee` package](https://search.nixos.org/packages?show=lychee&type=packages&query=lychee).

> You may use Nix to reproducibly build documentation sites and other static sites. Some packages will install documentation in their `out` or `doc` outputs, or maybe you have dedicated package where you've made your static site reproducible by running a generator (like [Hugo](https://gohugo.io/), or [mdBook](https://rust-lang.github.io/mdBook/)) in a derivation.

Reviewer: ```suggestion
You may use Nix to reproducibly build documentation sites and other static sites.
Some packages will install documentation in their `out` or `doc` outputs, or maybe you have dedicated package where you've made your static site reproducible by running a generator, such as [Hugo](https://gohugo.io/) or [mdBook](https://rust-lang.github.io/mdBook/), in a derivation.
```

> If you have a static site that is buildable in Nix, you can use `lycheeLinkCheck` to check that the hyperlinks in your site are correct, and do so as part of your Nix workflow and CI.

Reviewer: ```suggestion
If you have a static site that can be built with Nix, you can use `lycheeLinkCheck` to check that the hyperlinks in your site are correct, and do so as part of your Nix workflow and CI.
```

:::{.example #ex-lycheelinkcheck}

# Check hyperlinks in the `nix` documentation

```nix
testers.lycheeLinkCheck {
  site = nix.doc + "/share/doc/nix/manual";
}
```

:::

### Return value {#tester-lycheeLinkCheck-return}

This tester produces a package that does not produce useful outputs, but only succeeds if the hyperlinks in your site are correct. The build log will list the broken links.

> You may use it in two modes:

Reviewer: ```suggestion
It has two modes:
```

- Use the returned package directly, and the build process will check that internal hyperlinks are correct. This runs in the sandbox, so it will not check external hyperlinks, but it is quick and reliable.

> - Invoke the `.online` with [`nix run`](https://nixos.org/manual/nix/stable/command-ref/new-cli/nix3-run) ([experimental](https://nixos.org/manual/nix/stable/contributing/experimental-features#xp-feature-nix-command)). This runs outside the sandbox, and check that bith internal and external hyperlinks are correct. Example:

Reviewer: ```suggestion
- Invoke the `.online` attribute with [`nix run`](https://nixos.org/manual/nix/stable/command-ref/new-cli/nix3-run) ([experimental](https://nixos.org/manual/nix/stable/contributing/experimental-features#xp-feature-nix-command)). This runs outside the sandbox, and check that both internal and external hyperlinks are correct.
  Example:
```

  ```shell
  nix run nixpkgs#lychee.tests.ok.online
  ```

> ### Parameters {#tester-lycheeLinkCheck-params}
> 
> #### `site` (required) {#tester-lycheeLinkCheck-param-site}
> 
> The path to the files to check.
> 
> #### `extraConfig` (optional) {#tester-lycheeLinkCheck-param-extraConfig}
> 
> Extra configuration to pass to `lychee` in its [config file](https://github.com/lycheeverse/lychee/blob/master/lychee.example.toml). It is automatically [translated](https://nixos.org/manual/nixos/stable/index.html#sec-settings-nix-representable) to TOML.
> 
> Example: `{ "include_verbatim" = true; }`
> 
> #### `lychee` (optional) {#tester-lycheeLinkCheck-param-lychee}
> 
> The `lychee` package to use.
> 
> #### `remapUrl` (optional) {#tester-lycheeLinkCheck-param-remapUrl}
> 
> A URL to which the files may be deployed.
> Instead of ignoring this URL because the check is offline, `lychee` will check that the links to this URL are correct, by pretending `site` is deployed there.
> 

Reviewer: Per [conventions](https://github.com/NixOS/nixpkgs/blob/master/doc/README.md#documentation-conventions) we now use definition lists for parameter listings:

```suggestion
### Inputs {#tester-lycheeLinkCheck-inputs}

`site` (path or derivation) {#tester-lycheeLinkCheck-param-site}

: The path to the files to check.

`extraConfig` (attribute set) {#tester-lycheeLinkCheck-param-extraConfig}

: Extra configuration to pass to `lychee` in its [configuration file](https://github.com/lycheeverse/lychee/blob/master/lychee.example.toml).
  It is automatically [translated](https://nixos.org/manual/nixos/stable/index.html#sec-settings-nix-representable) to TOML.

  Example: `{ "include_verbatim" = true; }`

`lychee` (derivation, optional) {#tester-lycheeLinkCheck-param-lychee}

: The `lychee` package to use.
```

:::

## `lycheeLinkCheck` {#tester-lycheeLinkCheck}

Check a packaged static site's links with the [`lychee` package](https://search.nixos.org/packages?show=lychee&type=packages&query=lychee).

> You may use Nix to reproducibly build documentation sites and other static sites.

Reviewer: ```suggestion
You may use Nix to reproducibly build static websites, such as for software documentation.
```

  site = nix.doc + "/share/doc/nix/manual";
}
```

:::

### Return value {#tester-lycheeLinkCheck-return}

This tester produces a package that does not produce useful outputs, but only succeeds if the hyperlinks in your site are correct. The build log will list the broken links.

It has two modes:

- Build the returned derivation; its build process will check that internal hyperlinks are correct. This runs in the sandbox, so it will not check external hyperlinks, but it is quick and reliable.

- Invoke the `.online` attribute with [`nix run`](https://nixos.org/manual/nix/stable/command-ref/new-cli/nix3-run) ([experimental](https://nixos.org/manual/nix/stable/contributing/experimental-features#xp-feature-nix-command)). This runs outside the sandbox, and checks that both internal and external hyperlinks are correct.
  Example:

  ```shell
  nix run nixpkgs#lychee.tests.ok.online
  ```

### Inputs {#tester-lycheeLinkCheck-inputs}

`site` (path or derivation) {#tester-lycheeLinkCheck-param-site}

: The path to the files to check.

`remap` (attribe set, optional) {#tester-lycheeLinkCheck-param-remap}

: An attribute set where the attribute names are regular expressions.

>   The values should be strings, derivations or path values.

Reviewer: ```suggestion
  The values should be strings, derivations, or path values.
```

We seem to have converged on the Oxford comma. (Maybe I should write that into the guidelines...)

  In the returned check's default configuration, external URLs are only checked when you run the `.online` attribute.

>   By adding remappings, you can check offline that URLs to external resources are correct, by providing a file system based stand-in.

Reviewer: ```suggestion
  By adding remappings, you can check offline that URLs to external resources are correct, by providing a stand-in from the file system.
```

  Before checking the existence of a URL, the regular expressions are matched and replaced by their corresponding values.

  Example:

  ```nix
  {
    "https://nix\\.dev/manual/nix/[a-z0-9.-]*" = "${nix.doc}/share/doc/nix/manual";
    "https://nixos\\.org/manual/nix/(un)?stable" = "${emptyDirectory}/placeholder-to-disallow-old-nix-docs-urls";
  }
  ```

>   Store path in the attribute values are automatically prefixed with `file://`, because lychee requires this for paths in the file system.

Reviewer: ```suggestion
  Store paths in the attribute values are automatically prefixed with `file://`, because lychee requires this for paths in the file system.
```
